1-Draft/RFCNNNN-Implicit-Line-Continuance.md

```diff
@@ -0,0 +1,148 @@
+
+## Specification
+
+Note: This RFC is already implemented and submitted as [PR #9614](https://github.com/PowerShell/PowerShell/pull/9614).
+
+Since this RFC includes some breaking changes (see below), it will be initially implemented with the `PSImplicitLineContinuanceForNamedParameters` experimental feature flag. If the PowerShell Team and the community agree that the risk of this breaking change is low enough, and the workaround is sufficient for those low frequency use cases where it does become an issue, I would much prefer not using an experimental feature flag at all.
+
+For the implementation, this RFC builds on the work done in [PR #8938](https://github.com/PowerShell/PowerShell/pull/8938), evolving the logic used in the tokenizer that supports implicit line continuance for lines starting with a pipe such that it can be used to check for implicit line continuance for named parameters and splatted collections as well. To keep performance optimal, the parser logic that identifies command parameters where the implicit line continuance checks will be invoked will recognize if a pipe is found on a subsequent line as well, to ensure the lookahead logic is only invoked once per line in commands/pipelines.
+
+## Alternate Proposals and Considerations
+
+While implicit line continuance for splatted collections is not a breaking change (PowerShell syntax does not support `@variableName` at the start of a command), implicit line continuance for named parameters is a breaking change in some scenarios, because PowerShell syntax currently supports commands that start with dash, and PowerShell supports unary operators whose name starts with a single dash.
+
+For example, consider this script:
+
+```PowerShell
+function -dash {
+    'run'
+}
+
+Get-Process -Id $PID
+-dash
+```
+
+In PowerShell 6.x and earlier, two commands will be executed: `Get-Process`, and `-dash`. With this PR in place and the experimental feature enabled, only one command would be executed: `Get-Process`. The reason is that the parser would identify `-dash` as being intended as a parameter on the command on the previous line. That's how the parser needs to work to make this implicit line continuance functional.
+
+Similarly, consider this script:
+
+```PowerShell
+Get-Process -Id $PID
+-split 'a b c d'
+```
+
+Similar to the previous example, in PowerShell 6.x and earlier, one command and one statement will be executed: `Get-Process`, and the unary `-split` statement. With this PR in place and the experimental feature enabled, only one command would be executed: `Get-Process`, because again, the parser would identify `-split` as being intended as a parameter on the command on the previous line. The same applies to the `-join` unary operator as well. It does not apply to the `--` prefix arithmetic operator because the parser knows parameter names cannot start with a dash.
```

Comment:
In the initial implementation I didn't add support for the stop parsing sigil (`--%`), but I could easily add that. Right now the logic simply looks for something that parses as a parameter name or something that parses as a splatted collection. That means that non-PowerShell commands such as az.exe can have wrapping as well, e.g.

```PowerShell
az
    -h
```

Since the stop parsing sigil is not a valid parameter name, the lookahead logic identifies it as a separate command when used on the next line. I'm going to correct this such that lookahead identifies the stop parsing sigil as a continuance of the previous command, because that would be very useful with this as well. Thanks for the suggestion.

Reply:
This has been added to the RFC and the PR. Thanks again.